### ISSUE TITLE
[BUG] Unable to use forge/windows in arrayfire (MacOS Big Sur)

### ISSUE THREAD

**Issue by mfbolus** [arrayfire/arrayfire#3077]
Calls to forge like `graphics::ForgeManager::getMainWindow()` err for `Reason: image not found`. Specifically, it seems to be looking to open `/usr/local/opt/freetype/lib/libfreetype.6.dylib` which does not exist on my system running MacOS Big Sur (11.1). See below for more details re installation, etc.

Description
===========
* This apparent bug came up when I was trying to display an image; however, even compiled examples from `arrayfire/examples` directory give this error. 
* I installed ArrayFire v3.7.2 from prebuilt macOS package
* Have only tried CPU backend
* Currently have no workaround
* The bug is reproducible on my system (see below)
* I expected forge library to allow creation of windows, etc.
* Terminal output with AF_TRACE=all and AF_PRINT_ERRORS=1 environment
  variables set:
```
[platform][1610217884][102293] [ ../src/backend/common/DependencyModule.cpp:99 ] Attempting to load: libforge.dylib
[platform][1610217884][102293] [ ../src/backend/common/DependencyModule.cpp:104 ] Unable to open forge
ArrayFire v3.7.2 (CPU, 64-bit Mac OSX, build 218dd2c99)
In function fg_window graphics::ForgeManager::getMainWindow()
In file src/backend/common/graphics_common.cpp:274
Error loading Forge: dlopen(libforge.dylib, 1): Library not loaded: /usr/local/opt/freetype/lib/libfreetype.6.dylib
  Referenced from: /opt/arrayfire/lib/libforge.1.0.5.dylib
  Reason: image not found
Forge or one of it's dependencies failed to load. Try installing Forge or check if Forge is in the search path.
 0# graphics::ForgeManager::getMainWindow() in /opt/arrayfire/lib/libafcpu.3.7.2.dylib
 1# graphics::ForgeManager::getWindow(int, int, char const*, bool) in /opt/arrayfire/lib/libafcpu.3.7.2.dylib
 2# af_create_window in /opt/arrayfire/lib/libafcpu.3.7.2.dylib
 3# af::Window::initWindow(int, int, char const*) in /opt/arrayfire/lib/libafcpu.3.7.2.dylib
 4# main in /Users/mfbolus/Desktop/temp_build/graphics/./plot2d_cpu

ArrayFire Exception (Failed to load dynamic library. :501):
In function fg_window graphics::ForgeManager::getMainWindow()
In file src/backend/common/graphics_common.cpp:274
Error loading Forge: dlopen(libforge.dylib, 1): Library not loaded: /usr/local/opt/freetype/lib/libfreetype.6.dylib
  Referenced from: /opt/arrayfire/lib/libforge.1.0.5.dylib
  Reason: image not found
Forge or one of it's dependencies failed to load. Try installing Forge or check if Forge is in the search path.
 0# graphics::ForgeManager::getMainWindow() in /opt/arrayfire/lib/libafcpu.3.7.2.dylib
 1# graphics::ForgeManager::getWindow(int, int, char const*, bool) in /opt/arrayfire/lib/libafcpu.3.7.2.dylib
 2# af_create_window in /opt/arrayfire/lib/libafcpu.3.7.2.dylib
 3# af::Window::initWindow(int, int, char const*) in /opt/arrayfire/lib/libafcpu.3.7.2.dylib
 4# main in /Users/mfbolus/Desktop/temp_build/graphics/./plot2d_cpu

In function void af::Window::initWindow(const int, const int, const char *const)
In file src/api/cpp/graphics.cpp
libc++abi.dylib: terminating with uncaught exception of type af::exception: ArrayFire Exception (Failed to load dynamic library. :501):
In function fg_window graphics::ForgeManager::getMainWindow()
In file src/backend/common/graphics_common.cpp:274
Error loading Forge: dlopen(libforge.dylib, 1): Library not loaded: /usr/local/opt/freetype/lib/libfreetype.6.dylib
  Referenced from: /opt/arrayfire/lib/libforge.1.0.5.dylib
  Reason: image not found
Forge or one of it's dependencies failed to load. Try installing Forge or check if Forge is in the search path.
 0# graphics::ForgeManager::getMainWindow() in /opt/arrayfire/lib/libafcpu.3.7.2.dylib
 1# graphics::ForgeManager::getWindow(int, int, char const*, bool) in /opt/arrayfire/lib/libafcpu.3.7.2.dylib
 2# af_create_window in /opt/arrayfire/lib/libafcpu.3.7.2.dylib
 3# af::Window::initWindow(int, int, char const*) in /opt/arrayfire/lib/libafcpu.3.7.2.dylib
 4# main in /Users/mfbolus/Desktop/temp_build/graphics/./plot2d_cpu

In function void af::Window::initWindow(const int, const int, const char *const)
In file src/api/cpp/graphics.cpp
[0] Intel: Intel(R) Core(TM) i7-8569U CPU @ 2.80GHzzsh: abort      AF_TRACE=all AF_PRINT_ERRORS=1 ./plot2d_cpu
```

Reproducible Code and/or Steps
------------------------------
I can reproduce the error by building arrayfire `examples` with `cmake` and running e.g. `./plot2d_cpu`.

System Information
------------------
1. ArrayFire 3.7.2
2. [0] Intel: Intel(R) Core(TM) i7-8569U CPU @ 2.80GHzzsh
3. Compiler: Apple clang version 12.0.0 (clang-1200.0.32.28)

Checklist
---------
- [x] Using the latest available ArrayFire release
- [N/A] GPU drivers are up to date

**Comment by 9prady9** [comment]
@mfbolus On most Linux distributions `fontconfig` is available and since freetype is a direct dependency of it, both should be available. Not sure why that would change with any particular version of MacOS. I am not exactly a regular use of OSX perhaps @umar456 can shed some light onto that.

As a workaround, Can you please try installing fontconfig `brew install fontconfig` that should take care of it.

**Comment by mfbolus** [comment]
@9prady9 , thank you for the quick reply! I have installed fontconfig/freetype using macports (what I use for all such installations, for better or worse). This defaults to installing at `/opt/local`. I have `/opt/local/lib` in `DYLD_LIBRARY_PATH` already, but I will try building from source and circle back. It might be a couple days before I can get around to it, but I am fairly sure this will be the workaround. Maybe @umar456 has seen similar issue (having to install this dependency via brew) and it's not idiosyncratic to my system.

> While using ArrayFire, this tutorial might be of help. Just sharing in case you haven't already found it.

Had already run across the above, but thank you for the helpful hint!

Thanks again for getting back to me and for your great work.

**Comment by mfbolus** [comment]
@9prady9 , I have since compiled from source and can run examples like `plot2d`. One thing that I came across was the need to use the AF_BUILD_FORGE option for configuration. I did not see this option listed [here](https://github.com/arrayfire/arrayfire/wiki/Build-Instructions-for-OSX), so I had to hunt down why the forge library wasn't being built. I still cannot run a examples that depend on freeimage, but this is an issue building freeimage on MacOS 11. I am confident the cpu ArrayFire examples would run if I had this separate issue resolved.

In summary:
- The pre-built AF installation does not appear to work in MacOS Big Sur (11.2) for reasons in first post
- Building from source resolves this problem, but must remember to turn on AF_BUILD_FORGE option at configure

**Comment by 9prady9** [comment]
@mfbolus Forge is not a link time dependency of ArrayFire, only it's headers are sufficient to add graphics support to ArrayFire. Hence, the option you mentioned is turned off by default. Having said that it is not an advanced option although there is a chance you missed it if you didn't use `ccmake` or `cmake-gui`. Sorry about the wiki page not reflecting the `AF_BUILD_FORGE` option. I have updated the wiki page now.

@umar456 Can you please work with the @mfbolus on why pre-built ArrayFire binary's graphics dependencies are failed to be found.

**Comment by mfbolus** [comment]
@9prady9, I didn't mean to insinuate poor documentation or anything, but I appreciate the change to the wiki. As you guessed, I did not use `ccmake`.

**Comment by 9prady9** [comment]
> @9prady9, I didn't mean to insinuate poor documentation or anything, but I appreciate the change to the wiki. As you guessed, I did not use `ccmake`.

Sorry if my response gave such intent, I didn't mean to. It is good to fix and add missing documentation. As much as we strive to give best documentation we can, mistakes do happen and we have to fix them. Thank you for pointing it out!

**Comment by mfbolus** [comment]
Following up here to make sure I am not keeping something open that does not need to be. Building/installing from source works fine for my purposes. If no other users report an issue using Forge with the 3.7.2 macOS installer on Big Sur, then perhaps we should close this issue. Let me know.

**Comment by 9prady9** [comment]
@mfbolus We will look into it while generating latest stable release macOS installer. Lets have this open until @umar456 shares his input once he looks into it.

**Comment by lu4** [comment]
The official build steps for Forge from it's wiki (https://github.com/arrayfire/forge/wiki#osx):

```
cd /path/to/dir/forge
mkdir build && cd build && cmake ..
make -j8 # or number of threads of your choosing
```

is lacking a final step:

```
make install
```
Just make sure you are inside `build` folder before running above command. I was also blindly following build steps and for no apparent reason was expecting it to work, it's obvious that we need to install library if we expect it to be located by ArrayFire

By the way guys (ArrayFire creators) thank you for putting together such an awesome lib, the more I use it the more I like it.

**Comment by 9prady9** [comment]
@lu4 There is already section after this build commands section https://github.com/arrayfire/forge/wiki#installation Am I missing something here ?
